JEN-1012 5.6

abychko opened the pull request:
LD_PRELOAD for ASAN builds

[*] reorganised shell script to properly set LD_PRELOAD and MYSQLD_ENV

abychko committed:
JEN-1012
LD_PRELOAD for ASAN builds

[*] reorganised shell script to properly set LD_PRELOAD and MYSQLD_ENV

abychko committed:
JEN-1012

added actual ASAN library to LD_PRELOAD

abychko committed:
JEN-1012

adjusted libhotbackup testing

abychko committed:
JEN-1012

fixed libhotbackup test

abychko committed:
JEN-1012

adjusted tokudb test

abychko commented:
without ASAN https://ps2.cd.percona.com/job/test-percona-server-5.6-pipeline/21/console

abychko committed:
JEN-1012

removed PS directory before tar unpacking

abychko committed:
JEN-1012

trying to find right libasan from ldd output

abychko committed:
JEN-1012

used mysqld_env for ld_preload everywhere

abychko commented:
finally fixed ASAN logic https://ps2.cd.percona.com/job/test-percona-server-5.6-pipeline/25/consoleFull

abychko commented:
ready for review

laurynas-biveinis commented on local/test-binary:
```diff
@@ -15,14 +15,26 @@ set -o errexit
 set -o xtrace
 
 WORKDIR_ABS=$(cd ${1:-./build}; pwd -P)
+rm -fr ${WORKDIR_ABS}/PS
```
Why? Does this fix some other JEN?

abychko replied:
it fixes the following  issue: if test is running on the same node after build, then 3 mysqld binaries may be found in WORKDIR_ABS - from build tree, from installation directory and from tarball

laurynas-biveinis replied:
It would be more robust to delete all the non-needed copies then?
This would also help with find | head -1 below (both faster and removes the need for head -1).
Also how do we know that head -1 there picks the right copy?

laurynas-biveinis commented on local/test-binary:
```diff
@@ -15,14 +15,26 @@ set -o errexit
 cd ${WORKDIR_ABS}/PS/mysql-test
-
+#
+# CentOS 6 & 7
```
Why?

abychko replied:
to set paths before any actual calls

laurynas-biveinis commented on local/test-binary:
```diff
@@ -15,14 +15,26 @@ set -o errexit
 TOKUDB_PLUGIN=$(find $WORKDIR_ABS -type f -name 'ha_tokudb.so')
-HOTBACKUP_LIB=$(find $WORKDIR_ABS -type f -name 'libHotBackup.so')
-HOTBACKUP_PLUGIN=$(find $WORKDIR_ABS -type f -name 'tokudb_backup.so')
+HOTBACKUP_LIB=$(find $WORKDIR_ABS -type f -name 'libHotBackup.so' | head -n1)
```
Why?

abychko replied:
there are multiple binaries if build ran on the same node, please see above

abychko replied:
fixed, searching in PS directory after un-taring

laurynas-biveinis commented on local/test-binary:
```diff
@@ -15,14 +15,26 @@ set -o errexit
-HOTBACKUP_LIB=$(find $WORKDIR_ABS -type f -name 'libHotBackup.so')
-HOTBACKUP_PLUGIN=$(find $WORKDIR_ABS -type f -name 'tokudb_backup.so')
+HOTBACKUP_LIB=$(find $WORKDIR_ABS -type f -name 'libHotBackup.so' | head -n1)
+HOTBACKUP_PLUGIN=$(find $WORKDIR_ABS -type f -name 'tokudb_backup.so'| head -n1)
```
Why?

abychko replied:
pls see above

laurynas-biveinis replied:
See my reply above, this looks almost-broken

abychko replied:
currently it's 
tar -C ${WORKDIR_ABS}/PS --strip-components=1 -zxpf $(ls $WORKDIR_ABS/*.tar.gz | head -1)
WORKDIR_ABS contains all files, even garbage, but ${WORKDIR_ABS}/PS contains only unpacked tarball

abychko replied:
is not needed anymore

laurynas-biveinis commented on local/test-binary:
```diff
@@ -79,11 +88,12 @@ status=0
 # Running MTR test cases
 if [[ "${DEFAULT_TESTING}" != "no" ]]; then
-	LD_PRELOAD="${EATMYDATA}" MTR_BUILD_THREAD=auto ./mysql-test-run.pl \
+  MYSQLD_ENV="${ADD_TO_LD_PRELOAD:-} ${EATMYDATA}"
```
Why both here and in --mysqld-env?

abychko replied:
sorry?

laurynas-biveinis replied:
Sorry, misread it, it's good

laurynas-biveinis commented on local/test-binary:
```diff
@@ -93,45 +103,42 @@ if [[ "${DEFAULT_TESTING}" != "no" ]]; then
 fi
 
-
+# we test with system jemalloc
```
Why?

abychko replied:
because we need jemalloc only for tokudb tests and not for regular tests

laurynas-biveinis replied:
No harm (and potential benefit) to enable jemalloc for regular tests too

abychko replied:
added

laurynas-biveinis commented on local/test-binary:
```diff
@@ -93,45 +103,42 @@ if [[ "${DEFAULT_TESTING}" != "no" ]]; then
+  [[ -e lib/mysql/plugin/tokudb_backup.so ]] || ln -s ${HOTBACKUP_PLUGIN} lib/mysql/plugin/tokudb_backup.so
 
-    LD_PRELOAD="${EATMYDATA} ${JEMALLOC}" MTR_BUILD_THREAD=auto \
+    MYSQLD_ENV="${ADD_TO_LD_PRELOAD:-} ${JEMALLOC} ${EATMYDATA}"
```
Why both here and in --mysqld-env?

abychko replied:
Outdated

laurynas-biveinis reviewed (changes requested):
The test build
1) Does not exercise ASAN, hence invalid
2) Has `+ LD_PRELOAD=' /usr/local/lib/libeatmydata.so'`
Does libeatmydata actually get preloaded with this extra space before /usr?

laurynas-biveinis commented:
Sorry missed that there are two test runs. The first one is OK as ASAN-less test build.
The second one looks that it works, one suggestion:
`--mysqld-env=LD_PRELOAD=/lib64/libasan.so.4 /usr/local/lib/libeatmydata.so`
consider using colon not space as the separator

abychko committed:
JEN-1012

fixed LD_PRELOAD syntax

abychko committed:
JEN-1012

added jemalloc for regular builds

abychko committed:
JEN-1012

looking for toku* plugins and libs in unpacked directory

abychko commented:
+ MYSQLD_ENV=/lib64/libasan.so.4:/usr/lib64/libjemalloc.so.1:/usr/local/lib/libeatmydata.so
https://ps2.cd.percona.com/job/test-percona-server-5.6-pipeline/26/console

abychko commented:
hope it's last run https://ps2.cd.percona.com/job/test-percona-server-5.6-pipeline/26/console

laurynas-biveinis commented on local/test-binary:
```diff
@@ -52,25 +69,22 @@ function process_mtr_output {
-        exit 1
-    fi
+    # trying to find the proper library
+    MYSQLD=$(find ${WORKDIR_ABS}/PS -name mysqld | head -n1)
```
Why head -n1 here?

abychko replied:
not needed, removed

laurynas-biveinis commented on local/test-binary:
```diff
@@ -92,46 +107,37 @@ if [[ "${DEFAULT_TESTING}" != "no" ]]; then
-    mkdir -p lib/mysql/plugin
-    ln -s ${TOKUDB_PLUGIN}    lib/mysql/plugin/ha_tokudb.so
-    ln -s ${HOTBACKUP_PLUGIN} lib/mysql/plugin/tokudb_backup.so
+  [[ -d lib/mysql/plugin ]] || mkdir -p lib/mysql/plugin
```
Why?

abychko replied:
if test is running on the same VM after build or test retry()'d- directory and links are already exist. it's just check - do not create directory if it exists, it's normal logic, nothing surprising

laurynas-biveinis replied:
What if it was a different version build on the same VM?

abychko replied:
note line 18, directory to unpack the tarball to will be recreated

abychko committed:
JEN-1012

deleted head -n1 from find command looking for mysqld

abychko committed:
JEN-1012

deleted checks for already existing dirs because script will start every
time with clean directory
